misc/multistream-select/src/protocol.rs

```diff
@@ -0,0 +1,486 @@
+// DEALINGS IN THE SOFTWARE.
+
+//! Multistream-select protocol messages an I/O operations for
+//! constructing protocol negotiation flows.
+//!
+//! A protocol negotiation flow is constructed by using the
+//! `Stream` and `Sink` implementations of `MessageIO` and
+//! `MessageReader`.
+
+use bytes::{Bytes, BytesMut, BufMut};
+use crate::length_delimited::{LengthDelimited, LengthDelimitedReader};
+use futures::{prelude::*, try_ready};
+use log::trace;
+use std::{io, fmt, error::Error, convert::TryFrom};
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint as uvi;
+
+/// The maximum number of supported protocols that can be processed.
+const MAX_PROTOCOLS: usize = 1000;
+
+/// The maximum length (in bytes) of a protocol name.
+///
+/// This limit is necessary in order to be able to unambiguously parse
+/// response messages without knowledge of the corresponding request.
+/// 140 comes about from 3 * 47 = 141, where 47 is the ascii/utf8
+/// encoding of the `/` character and an encoded protocol name is
+/// at least 3 bytes long (uvi-length followed by `/` and `\n`).
+/// Hence a protocol list response message with 47 protocols is at least
+/// 141 bytes long and thus such a response cannot be mistaken for a
+/// single protocol response. See `Message::decode`.
+const MAX_PROTOCOL_LEN: usize = 140;
+
+/// The encoded form of a multistream-select 1.0.0 header message.
+const MSG_MULTISTREAM_1_0: &[u8] = b"/multistream/1.0.0\n";
+/// The encoded form of a multistream-select 2.0.0 header message.
+const MSG_MULTISTREAM_2_0: &[u8] = b"/multistream/2.0.0\n";
+/// The encoded form of a multistream-select 'na' message.
+const MSG_PROTOCOL_NA: &[u8] = b"na\n";
+/// The encoded form of a multistream-select 'ls' message.
+const MSG_LS: &[u8] = b"ls\n";
+
+/// The known multistream-select protocol versions.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Version {
+    /// The first and currently still the only deployed version
+    /// of multistream-select.
+    V1,
+    /// Draft: https://github.com/libp2p/specs/pull/95
+    V2,
+}
+
+/// A protocol (name) exchanged during protocol negotiation.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Protocol(Bytes);
+
+impl AsRef<[u8]> for Protocol {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<Bytes> for Protocol {
+    type Error = ProtocolError;
+
+    fn try_from(value: Bytes) -> Result<Self, Self::Error> {
+        if !value.as_ref().starts_with(b"/") || value.len() > MAX_PROTOCOL_LEN {
+            return Err(ProtocolError::InvalidProtocol)
+        }
+        Ok(Protocol(value))
+    }
+}
+
+impl TryFrom<&[u8]> for Protocol {
+    type Error = ProtocolError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        Self::try_from(Bytes::from(value))
+    }
+}
+
+impl fmt::Display for Protocol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", String::from_utf8_lossy(&self.0))
```

Comment:
I think this is ok for now (although I don't like using the usage of the `Display` trait), but if multistream-select forces you to use UTF-8, I think we should open an issue to change the design to use `String` and `&str`. We need to look at the specs of multistream-select 2.0 first, however, in order to not have to potentially revert that later.